OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromCDLogic.cs

```diff
@@ -64,6 +80,50 @@ void CheckForDisk()
 			}
 		}
 
+		void InstallTFD(string source)
+		{
+			backButton.IsDisabled = () => true;
+			retryButton.IsDisabled = () => true;
+			insertDiskContainer.IsVisible = () => false;
+			installingContainer.IsVisible = () => true;
+			progressBar.Percentage = 0;
+
+			new Thread(() =>
+			{
+				using (var cab_ex = new InstallShieldCABExtractor(source))
+				{
+					var denom = installData.IscfFileIds.Length + installData.IscfPackageIds.Length;
+					var installPercent = 100 / denom;
+
+					foreach (uint index in installData.IscfFileIds)
+					{
+						var filename = cab_ex.FileName(index);
+						var dest = Platform.ResolvePath("^", "Content", Game.ModData.Manifest.Mod.Id, filename.ToLower());
+						cab_ex.ExtractFile(index, dest);
+						progressBar.Percentage += installPercent;
+					}
+
+					var ArchivesToExtract = installData.IscfPackageIds.Select(x => x.Split(':'));
+					foreach (var archive in ArchivesToExtract)
+					{
+						var filename = cab_ex.FileName(uint.Parse(archive[0]));
+						var destFile = Platform.ResolvePath("^", "Content", Game.ModData.Manifest.Mod.Id, filename.ToLower());
+						cab_ex.ExtractFile(uint.Parse(archive[0]), destFile);
+
+						var annotation = archive.Length > 1 ? archive[1] : null;
+						var extractFiles = installData.ExtractFilesFromCD;
+						var onProgress = (Action<string>)(s => { });
```

Comment:
You can pull these outside the loop to stop them from being reallocated every pass.